elyra/pipeline/component_parameter.py

```diff
@@ -44,21 +49,85 @@
 from elyra.util.kubernetes import is_valid_label_value
 
 
-class ElyraProperty:
+class PropertyAttribute:
     """
-    A component property that is defined and processed by Elyra.
+    An attribute of an ElyraProperty instance that provides the means to construct the
+    schema for a property and contains information for processing property instances.
     """
 
+    _input_type_to_default_value = {"boolean": False, "array": "[]", "object": "{}", "string": ""}
+
+    def __init__(
+        self,
+        attribute_id: str,
+        display_name: Optional[str] = None,
+        placeholder: Optional[Any] = None,
+        default_value: Optional[Any] = None,
+        input_type: Optional[str] = None,
+        enum: Optional[List[Any]] = None,
+        hidden: Optional[bool] = False,
+        required: Optional[bool] = False,
+    ):
+        """
+        :param attribute_id: a shorthand id for this attribute, e.g. "env_var"
+        :param display_name: the display name for this attribute
+        :param placeholder: a placeholder value to be shown in the input field for this attribute
+        :param default_value: the default value to assign the attribute
+        :param input_type: the JSON data type of this attribute ("string", "boolean", "number", "array", or "object")
+        :param enum: a list of possible values that this attribute can take (will appear in a dropdown menu)
+        :param hidden: whether this attribute should be hidden in the UI, preventing users from entering a value
+        :param required: whether a value for this attribute is required
+        """
+        self.id = attribute_id
+        self.title = display_name
+        self.placeholder = placeholder
+        self.default_value = default_value or self._input_type_to_default_value.get(input_type)
+        self.input_type = input_type
+        self.enum = enum
+        self.hidden = hidden
+        self.required = required
+
+
+class ListItemPropertyAttribute(PropertyAttribute):
+    """
+    An attribute of an ElyraPropertyListItem instance that provides the means to construct the
+    schema for a property and contains information for processing property instances.
+    """
+
+    def __init__(
+        self,
+        attribute_id: str,
+        display_name: Optional[str] = None,
+        placeholder: Optional[Any] = None,
+        default_value: Optional[Any] = None,
+        input_type: Optional[str] = None,
+        enum: Optional[List[Any]] = None,
+        hidden: Optional[bool] = False,
+        required: Optional[bool] = False,
+        use_in_key: Optional[bool] = True,
+    ):
+        """
+        :param use_in_key: whether this attribute should be used when constructing a
+            key for an instance that will be used to de-duplicate list items
+        """
+        super().__init__(attribute_id, display_name, placeholder, default_value, input_type, enum, hidden, required)
+        self.use_in_key = use_in_key
+
+
+class ElyraProperty(ABC):
```

Comment:
Should this be also extending `EnforceOverrides` to raise errors as documented at https://pypi.org/project/overrides/
> Use `EnforceOverrides` to require subclass methods that shadow superclass methods to be decorated with @overrides.

```suggestion
class ElyraProperty(ABC, EnforceOverrides):
```

Reply:
per offline discussion, the commit to add `overrides` will be reverted and a new issue opened to investigate the addition later on